logger.go

```diff
@@ -244,9 +283,141 @@ func (logger *Logger) getHook(name string) *leveledHook {
 	return logger.hooks[name]
 }
 
+// AddError let functions to add error to the current logger indicating that something went
+// wrong in the logging process.
+func (logger *Logger) AddError(err error) {
+	if err != nil {
+		logger.errors = append(logger.errors, err)
+	}
+}
+
+// GetError returns the current error state of the logging process.
+func (logger *Logger) GetError() error { return logger.errors.AsError() }
+
+// ClearError cleans up the current error state of the logging process.
+// It also returns the current error state.
+func (logger *Logger) ClearError() error {
+	current := logger.errors.AsError()
+	logger.errors = nil
+	return current
+}
+
 // Writer is the implementation of io.Writer. You should not call directly this function.
 // The function will fail if called directly on a stream that have not been configured as out stream.
-func (logger *Logger) Write(p []byte) (n int, err error) {
+func (logger *Logger) oldWrite(p []byte) (n int, err error) {
 	logger.Print(string(p))
 	return len(p), nil
 }
+
+// Close implements io.Closer
+func (logger *Logger) Close() error {
+	if logger.remaining != "" {
+		_, err := logger.Write(nil)
+		return err
+	}
+	return nil
+}
+
+// This methods intercepts every message written to stream if Catcher is set and determines if a logging
+// function should be used.
+func (logger *Logger) Write(writeBuffer []byte) (int, error) {
+	if !logger.Catcher {
+		if err := logger.printLines(string(writeBuffer)); err != nil {
+			return 0, err
+		}
+		return len(writeBuffer), nil
+	}
+
+	var (
+		buffer      string
+		resultCount int
+	)
+
+	if logger.remaining != "" {
```

Comment:
Black magic below :smile:

Reply:
It's the same magic that was in terragrunt. I simply moved the code and adapt it a bit. The tests are pretty complete to ensure that this work in all cases. The problem is that we don't always control the order of the call to Write functions and some time, with a buffered output, lines can be cut in half.